wire-ios-utilities/Source/Public/SafeSubscript.swift

```
@@ -0,0 +1,26 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public extension Collection {
+    /// Returns the element at the specified index if it is within bounds, otherwise nil.
+    subscript(safe index: Index) -> Element? {
+        indices.contains(index) ? self[index] : nil
+    }
+}
```

issue: Personally I think this as an anti-pattern and am against adding it to our utilities and is one of the things that the Swift team currently refuses to add to the standard library. See discussion in this thread: https://forums.swift.org/t/safe-random-access-collection-element/36509/6

I wouldn't mind it being a private helper in a particular file but if it is to be something global I think it should be discussed in the architecture forum.

---

We check index if within bounds everywhere, why this subscript is different from just another if? It's just syntax sugar. 
If you use it wise and not overuse just for simplicity I do not see a problem in it. 

Apple has their own responsibility to trend smth or not because they have users of swift of all skill levels. Here in Wire we all pretty experienced not to overuse it in bad way I think

---

@samwyndham what are your reservations about it? From my understanding, adding a safe subscript is ignoring a larger problem which is not working with safe indices to begin with. @dmitrysimkin for what purpose do we need this exactly and why do we anticipate using invalid indices?

---

@johnxnguyen It is definitely a opinionated subject. These two comments from this thread express my opinion:

1. 
> You have an index into a collection from somewhere. This index has a meaning (maybe it's a particular element, or the beginning of a range, whatever).
> Now you use the safe accessor, ostensibly because you aren't sure the index is valid anymore -- so the collection may have changed, I guess? If you aren't sure whether the collection hasn't changed in the meantime, how do you know that the safe accessor will not return you a different element (now at the original index) than the one you thought you were dealing with?
> I fear this "safe" access pattern will lead to more code that is subtly wrong, but we'll notice it in fewer instance and thus have a harder time to fix it.

2. > The tl;dr here is that safe does not mean, "does not crash." It means "don't silently propagate invalid data", which can be accomplished here by either crashing or returning nil. As out of bounds indexing is a logic error, crashing is appropriate.

---

**suggestion**: could we use a different label instead of `"safe"`?
I don't mind having this helper as long as it's just a getter, although generally I prefer to write plain Swift code.
```suggestion
public extension Collection {
    /// Returns the element at the specified index if it is within bounds, otherwise nil.
    subscript(ifExists index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
```

---

I call it to get `let lastVisibleMessage = dataSource.allMessages[safe: firstIndexPath.section]` which may be not existing anymore when it was deleted or for self-deleting messages. 
Messages indeed can be updated as we have fetch result controller which is updated when there is a change in DB, so the safe check will skip for this time and on next reload (which can happen with delay because of debauncer) it will be in synced

---

@dmitrysimkin I'm just afraid of silent failures and this convenience can be an easy way to ignore issues. However, maybe this is just personal opinion, maybe everyone else likes it. I see that Chris Lattner is in favor of it from the linked thread above 😄 Feel free to ignore this comment

---

@samwyndham  I agree about ignoring issues, but I think here I have reasons to check if it's out of bounds. And we can try not to over use it.

@caldrian I've updated to `ifExists`